### PR TITLE
Add new manage_ec2_instance role

### PR DESCRIPTION
**Pull request by mandar242**
This adds a new role to create a ec2 instance with customizable configuration. 
See: https://issues.redhat.com/browse/ACA-1959

### CODE REVIEW

File: roles/ec2_instance_create/README.md
Comment (mandar242) [reply roles/ec2_instance_create/README.md:64]: Removed. Was added for relevant functionality, but not in scope of jira.

File: roles/ec2_instance_create/defaults/main.yml
Comment (mandar242) [reply roles/ec2_instance_create/defaults/main.yml:2]: added

File: roles/ec2_instance_create/tasks/main.yml
Comment (alinabuzachis) [inline roles/ec2_instance_create/tasks/main.yml:21]: What if a security group with this name already exist?

File: roles/ec2_instance_create/tasks/main.yml
Comment (alinabuzachis) [inline roles/ec2_instance_create/tasks/main.yml:28]: I would also assert changes where applicable to validate the behaviour.

File: roles/ec2_instance_create/tasks/main.yml
Comment (alinabuzachis) [inline roles/ec2_instance_create/tasks/main.yml:8]: What if an instance with this name already exist?

File: roles/ec2_instance_create/README.md
Comment (alinabuzachis) [inline roles/ec2_instance_create/README.md:61]: What if I'd like to associate and already existing sg?

**Review by alinabuzachis — COMMENTED** [review]
Since we are here, shouldn't we also provide a manual for this role?

File: roles/ec2_instance_create/README.md
Comment (mandar242) [reply roles/ec2_instance_create/README.md:3]: I am not sure on that, would like to defer to @GomathiselviS.
I can add the functionality if we decide to.

File: roles/ec2_instance_create/tasks/main.yml
Comment (mandar242) [reply roles/ec2_instance_create/tasks/main.yml:21]: updated the name to be more unique in general. Also `ec2_instance_create_external_sg_name` param is provided for users to specify the sg name.
Additionally, added new param `ec2_instance_create_external_sg` which user can provide to associate instance to existing external sg.

File: roles/ec2_instance_create/tasks/main.yml
Comment (mandar242) [reply roles/ec2_instance_create/tasks/main.yml:8]: `ec2_instance_create_instance_name` param can be used by users to provide a name of their choice.

File: roles/ec2_instance_create/README.md
Comment (mandar242) [reply roles/ec2_instance_create/README.md:64]: added the functionality https://github.com/redhat-cop/cloud.aws_ops/pull/123/files#diff-88ebaff2a93da854e7c0b2b685050632c51e2530780423cb598a0864aad05a3dR14

File: roles/ec2_instance_create/tasks/main.yml
Comment (GomathiselviS) [reply roles/ec2_instance_create/tasks/main.yml:22]: Hi @alinabuzachis, thank you for all your suggestions and they are all valid. This role has been added to enable automation experiences for the cloud content and has a hard deadline, with the specific use case already defined and documented. I'd like to add all these functionalities to the role in a separate PR. @mandar242, we can still include the delete functionality in this PR. If either of you could create an issue to track these points, it would be really helpful.

Comment (mandar242) [reply roles/ec2_instance_create/tasks/main.yml:22]: @GomathiselviS @alinabuzachis 
sounds good. I'll keep track of these suggestions, and can work on adding those once this PR is merged.
For now, I'll work on adding the ability to delete instance and other resources created by the role.

File: roles/ec2_instance_create/tasks/main.yml
Comment (mandar242) [reply roles/ec2_instance_create/tasks/main.yml:23]: added a check, playbook will fail if instance with same name exists in the region.

File: roles/ec2_instance_create/README.md
Comment (gravesm) [inline roles/ec2_instance_create/README.md:68]: Does this replace or add to the security group defined in `ec2_instance_create_external_sg_id`? It might be helpful to clarify what the relationship is between these two security groups.

Comment (mandar242) [reply roles/ec2_instance_create/README.md:75]: updated readme.

File: roles/ec2_instance_create/tasks/ec2_instance_create_operations.yml
Comment (hakbailey) [inline roles/ec2_instance_create/tasks/ec2_instance_create_operations.yml:10]:
```suggestion
- name: Verify that an instance with same name does not exist
```

File: roles/ec2_instance_create/tasks/ec2_instance_create_operations.yml
Comment (hakbailey) [inline roles/ec2_instance_create/tasks/ec2_instance_create_operations.yml:44]: Are these two defaults needed? It looks like they're already defined in the role defaults.

File: roles/ec2_instance_create/tasks/ec2_instance_create_operations.yml
Comment (hakbailey) [inline roles/ec2_instance_create/tasks/ec2_instance_create_operations.yml:56]: Same here.

File: roles/ec2_instance_create/README.md
Comment (mandar242) [reply roles/ec2_instance_create/README.md:33]: updated code and added note that key pair will be created with provided name if a key does not exists with the given name in the given region

File: roles/ec2_instance_create/tasks/ec2_instance_delete_operations.yml
Comment (mandar242) [reply roles/ec2_instance_create/tasks/ec2_instance_delete_operations.yml:4]: fixed. can be deleted only with instance name and region as well.

File: roles/ec2_instance_create/README.md
Comment (mandar242) [reply roles/ec2_instance_create/README.md:72]: Was added as it was suggested in PR review comments. I can revert and remove `external` keyword if everyone okay with that.
cc @GomathiselviS @alinabuzachis

File: roles/ec2_instance_create/README.md
Comment (mandar242) [reply roles/ec2_instance_create/README.md:33]: Changed `ec2_instance_create_key_name` to `optional`

File: roles/ec2_instance_create/README.md
Comment (mandar242) [reply roles/ec2_instance_create/README.md:37]: changed to `optional`

File: roles/ec2_instance_create/README.md
Comment (mandar242) [reply roles/ec2_instance_create/README.md:50]: I think it would useful having it around for additional flexibility to users. But fine with removing if everyone in agreement. 
cc @alinabuzachis @GomathiselviS

File: roles/ec2_instance_create/README.md
Comment (mandar242) [reply roles/ec2_instance_create/README.md:15]: updated.

File: roles/ec2_instance_create/meta/argument_specs.yml
Comment (mandar242) [reply roles/ec2_instance_create/meta/argument_specs.yml:16]: renamed.

File: roles/ec2_instance_create/tasks/ec2_instance_create_operations.yml
Comment (mandar242) [reply roles/ec2_instance_create/tasks/ec2_instance_create_operations.yml:9]: fixed.

File: roles/ec2_instance_create/README.md
Comment (GomathiselviS) [reply roles/ec2_instance_create/README.md]: It was suggested based on the requirements outlined in [ACA-1959](https://issues.redhat.com/browse/ACA-1959).

File: roles/ec2_instance_create_delete/README.md
Comment (hakbailey) [inline roles/ec2_instance_create_delete/README.md:25]: The choices listed here don't match the arg spec. I think we probably want to stick with just create and delete for this use case?

File: roles/ec2_instance_create_delete/README.md
Comment (hakbailey) [inline roles/ec2_instance_create_delete/README.md:48]:
```suggestion
  If not provided, instance will be created in the default subnet for the default vpc in the AWS region if present.
```

File: roles/ec2_instance_create_delete/meta/argument_specs.yml
Comment (hakbailey) [inline roles/ec2_instance_create_delete/meta/argument_specs.yml:54]:
```suggestion
          - Whether to wait for the EC2 instance to be in the running or terminated state before continuing.
```

File: roles/ec2_instance_create_delete/README.md
Comment (hakbailey) [inline roles/ec2_instance_create_delete/README.md:60]: I'm not entirely sure why this is needed. Shouldn't we just use the VPC of the instance?

File: roles/ec2_instance_create_delete/README.md
Comment (hakbailey) [inline roles/ec2_instance_create_delete/README.md:84]:
```suggestion
  Whether to create and associate a security group with the EC2 instance. Default is `false`.
  If set to `true`, a security group will be created or associated with the instance.
```

File: roles/ec2_instance_create_delete/README.md
Comment (hakbailey) [inline roles/ec2_instance_create_delete/README.md:89]: Why not default to something like `{instance-name}-external-sg`?

File: roles/ec2_instance_create_delete/tasks/ec2_instance_delete_operations.yml
Comment (hakbailey) [inline roles/ec2_instance_create_delete/tasks/ec2_instance_delete_operations.yml:12]: Should we also delete the keypair?

File: roles/ec2_instance_create_delete/meta/argument_specs.yml
Comment (hakbailey) [inline roles/ec2_instance_create_delete/meta/argument_specs.yml:32]: If these are required here then we can't delete the instance without providing them, which I don't think should be necessary.

File: roles/ec2_instance_create/README.md
Comment (hakbailey) [reply roles/ec2_instance_create/README.md]: Ha, I just started to add a similar comment as Aubin, we should just use the VPC of the ec2 instance.

File: roles/ec2_instance_create/README.md
Comment (hakbailey) [reply roles/ec2_instance_create/README.md]: I left a comment about this but I disagree, for this use case I think we should stick to create and delete.

File: roles/ec2_instance_create_delete/README.md
Comment (hakbailey) [inline roles/ec2_instance_create_delete/README.md:110]:
```suggestion
    - role: cloud.aws_ops.ec2_instance_create_delete
```

File: roles/ec2_instance_create_delete/tasks/ec2_instance_create_operations.yml
Comment (hakbailey) [inline roles/ec2_instance_create_delete/tasks/ec2_instance_create_operations.yml:4]:
```suggestion
    - name: Get instance info with provided name
```

File: roles/ec2_instance_create_delete/tasks/ec2_instance_create_operations.yml
Comment (hakbailey) [inline roles/ec2_instance_create_delete/tasks/ec2_instance_create_operations.yml:18]:
```suggestion
  when: ec2_instance_create_delete_key_name is defined and ec2_instance_create_delete_key_name | length > 0
```

**Review by hakbailey — CHANGES_REQUESTED** [review]

File: roles/ec2_instance_create_delete/README.md
Comment (hakbailey) [inline roles/ec2_instance_create_delete/README.md:95]: Let's not default to this. Probably best to just not include any inbound rules by default?

File: roles/ec2_instance_create_delete/README.md
Comment (GomathiselviS) [inline roles/ec2_instance_create_delete/README.md:87]:
```suggestion
    - role: cloud.aws_ops.ec2_instance_create_delete
```

File: roles/ec2_instance_create_delete/README.md
Comment (GomathiselviS) [inline roles/ec2_instance_create_delete/README.md:89]:
```suggestion
          ec2_instance_create_delete_operation: create
```

File: roles/ec2_instance_create_delete/README.md
Comment (GomathiselviS) [inline roles/ec2_instance_create_delete/README.md:117]:
```suggestion
    - role: cloud.aws_ops.ec2_instance_create_delete
```

File: roles/ec2_instance_create_delete/README.md
Comment (GomathiselviS) [inline roles/ec2_instance_create_delete/README.md:119]:
```suggestion
        ec2_instance_create_delete_operation: create
```

File: roles/ec2_instance_create_delete/meta/argument_specs.yml
Comment (GomathiselviS) [inline roles/ec2_instance_create_delete/meta/argument_specs.yml:24]:
```suggestion
          - The instance type for the EC2 instance. Required when `ec2_instance_create_delete_operation` is `create`.
```

File: roles/ec2_instance_create_delete/meta/argument_specs.yml
Comment (GomathiselviS) [inline roles/ec2_instance_create_delete/meta/argument_specs.yml:29]:
```suggestion
          - The AMI ID for the EC2 instance. Required when `ec2_instance_create_delete_operation` is `create`.
```

File: roles/ec2_instance_create_delete/README.md
Comment (GomathiselviS) [inline roles/ec2_instance_create_delete/README.md:1]:
```suggestion
# ec2_instance_create_delete
```

**Review by GomathiselviS — CHANGES_REQUESTED** [review]
Just a small suggestion, not a requirement. Do you think it would be better to rename the role to something like ec2_instance_management or manage_ec2_instance, or something along those lines?

File: roles/ec2_instance_create_delete/README.md
Comment (alinabuzachis) [reply roles/ec2_instance_create_delete/README.md:87]: I +1 for `manage_ec2_instance` and was thinking the same.

Comment (hakbailey) [reply roles/ec2_instance_create_delete/README.md:87]: Yep, I totally agree. Will update the role name.

Comment (hakbailey) [reply roles/ec2_instance_create_delete/README.md:87]: What do we think about `ec2_instance_provision`? The role doesn't really provide full instance management...

**Review by GomathiselviS — APPROVED** [review]

**Review by hakbailey — APPROVED** [review]

**Review by alinabuzachis — APPROVED** [review]

**Review by gravesm — APPROVED** [review]

File: roles/manage_ec2_instance/tasks/ec2_instance_create_operations.yml
Comment (abikouo) [inline roles/manage_ec2_instance/tasks/ec2_instance_create_operations.yml:50]:
```suggestion
```
has no effect when allocating address

**Review by abikouo — CHANGES_REQUESTED** [review]

File: roles/manage_ec2_instance/tasks/ec2_instance_create_operations.yml
Comment (hakbailey) [reply roles/manage_ec2_instance/tasks/ec2_instance_create_operations.yml:67]: Arguably we would also need to save the instance ip address. Since the private key is available via `new_key_pair_result.key.private_key` just like the ec2 instance info is available via `_ec2_instance.instances[0]`, I think it's fine to leave as is for now.

Comment (abikouo) [reply roles/manage_ec2_instance/tasks/ec2_instance_create_operations.yml:66]: If we leave it as it is, the role is unusable as there is no means to connect to the instance created (no private key file and IP address), the only use case is spawning an instance with a key pair and IP

Comment (hakbailey) [reply roles/manage_ec2_instance/tasks/ec2_instance_create_operations.yml:66]: The two vars I mentioned above are available to any playbook that uses this role. A user can reference them in the same way that we suggest they reference `ec2_networking_resources_subnet_result.subnet.id` when combining this role with the ec2_networking resources role. I can update the var names and the README to make that more clear, but we don't need to create new variables.

**Review by abikouo — APPROVED** [review]

File: roles/manage_ec2_instance/tasks/ec2_instance_create_operations.yml
Comment (hakbailey) [reply roles/manage_ec2_instance/tasks/ec2_instance_create_operations.yml]: Updated, I think it should be more clear now.

**Pull request merged**
